neqo-transport/src/connection/tests/stream.rs

```diff
@@ -618,3 +724,145 @@ fn increase_decrease_flow_control() {
     change_flow_control(StreamType::UniDi, RECV_BUFFER_NEW_SMALLER);
     change_flow_control(StreamType::BiDi, RECV_BUFFER_NEW_SMALLER);
 }
+
+#[test]
+fn session_flow_control_stop_sending_state_recv() {
+    const SMALL_MAX_DATA: usize = 1024;
+
+    let mut client = default_client();
+    let mut server = new_server(
+        ConnectionParameters::default().max_data(u64::try_from(SMALL_MAX_DATA).unwrap()),
+    );
+
+    connect(&mut client, &mut server);
+
+    let stream_id = client.stream_create(StreamType::UniDi).unwrap();
+    assert_eq!(
+        client.stream_avail_send_space(stream_id).unwrap(),
+        SMALL_MAX_DATA
+    );
+
+    // send 1 byte so that the server learns about the stream.
+    assert_eq!(
+        client
+            .stream_send(stream_id, &vec![b'a'].into_boxed_slice())
+            .unwrap(),
+        1
+    );
+
+    exchange_data(&mut client, &mut server);
+
+    server
+        .stream_stop_sending(stream_id, Error::NoError.code())
+        .unwrap();
+
+    assert_eq!(
+        client
+            .stream_send(stream_id, &vec![b'a'; RECV_BUFFER_SIZE].into_boxed_slice())
+            .unwrap(),
+        SMALL_MAX_DATA - 1
+    );
+
+    exchange_data(&mut client, &mut server);
+
+    // The flow control should have been updated and the clinet caan again send
+    // SMALL_MAX_DATA.
+    let stream_id2 = client.stream_create(StreamType::UniDi).unwrap();
+    assert_eq!(
+        client.stream_avail_send_space(stream_id2).unwrap(),
+        SMALL_MAX_DATA
+    );
+}
+
+#[test]
+fn session_flow_control_stop_sending_state_size_known() {
+    const SMALL_MAX_DATA: usize = 1024;
+
+    let mut client = default_client();
+    let mut server = new_server(
+        ConnectionParameters::default().max_data(u64::try_from(SMALL_MAX_DATA).unwrap()),
+    );
+
+    connect(&mut client, &mut server);
+
+    let stream_id = client.stream_create(StreamType::UniDi).unwrap();
+    assert_eq!(
+        client.stream_avail_send_space(stream_id).unwrap(),
+        SMALL_MAX_DATA
+    );
+
+    // send 1 byte so that the server learns about the stream.
+    assert_eq!(
+        client
+            .stream_send(stream_id, &vec![b'a'; RECV_BUFFER_SIZE].into_boxed_slice())
+            .unwrap(),
+        SMALL_MAX_DATA
+    );
+
+    let out1 = client.process(None, now()).dgram();
+    // Delay his packet and lett the server receive fin first (it will enter SizeKnown state).
+    client.stream_close_send(stream_id).unwrap();
+    let out2 = client.process(None, now()).dgram();
+
+    mem::drop(server.process(out2, now()).dgram());
+
+    server
+        .stream_stop_sending(stream_id, Error::NoError.code())
+        .unwrap();
+
+    let out = server.process(out1, now()).dgram();
+    let _out = client.process(out, now()).dgram();
+
+    // The flow control should have been updated and the clinet caan again send
+    // SMALL_MAX_DATA.
```

Comment:
session_flow_control_stop_sending_state_recv tests the case when we need to wait for the reset. Because of the way the frames are send the state transfers from AbortReading -> WaitingForReset(while processing acks) -> Recvd (when reset is processed). In this test we call exchange_packets because we need the reset frame. MAX_DATA is send after RESET is received.

session_flow_control_stop_sending_state_size_known tests the case where we have the final size and we do not need a RESET frame. Here already sop_sending() call releases flow control and MAX_DATA is send together with STOP_SENDING frame.

I will add some comments to the test.